### PR TITLE
Attn : Dylan :: Malformed Cookie Bug : 17 December 2015

### PR DESCRIPTION
**Pull request by swashcap**
Asana task: https://app.asana.com/0/37131652076106/74571845162511

### CODE REVIEW

File: test/auth.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
+/**
+ * Unit tests for the logon widget's 'Auth' service.
+ *
+ * @todo  Move login tests to this file.
+ */
+'use strict';
+
+var Auth = require('../scripts/lib/auth');
+var Cookies = require('js-cookie');
```
Comment (cdaringe) [inline test/auth.js]: are `Auth` and `Cookies` constructors?

Comment (swashcap) [inline test/auth.js]: js-cookie uses [`Cookies` in their docs](https://github.com/js-cookie/js-cookie#basic-usage). Not sure why I capitalized `Auth`. I can change it.

Comment (swashcap) [inline test/auth.js]: Fixed in 7d22647c4e20ce8eb2e9a375ff1cb37768c32319.

File: scripts/lib/auth.js [hunk 2]
```diff
@@ -64,23 +80,36 @@ function getApiUrl(endpoint) {
...
-        domain,
-        ';'
-    ].join('');
+    var authCookieName = getOptions().authCookieName;
+    var hostPieces = location.hostname.split('.');
+    var options = {
+        path: '/'
+    };
+
+    if (hostPieces.length > 2) {
```
Comment (dylancwood) [inline scripts/lib/auth.js:97]: Isn't the domain always the last two `hostPieces`, regardless of length?

File: scripts/lib/auth.js [hunk 3]
```diff
@@ -64,23 +80,36 @@ function getApiUrl(endpoint) {
...
+        options.domain = '.' + hostPieces.slice(hostPieces.length - 2).join('.');
```
Comment (dylancwood) [inline scripts/lib/auth.js]: You can just do hostPieces.slice(-2)` to grab the last two elements of an array.

proof:

```
> ['a', 'b', 'c'].slice(-2)
< ["b", "c"]
```

File: scripts/lib/auth.js [hunk 4]
```diff
@@ -167,9 +196,11 @@ function login(username, password) {
 /**
  * Log out.
  *
+ * @param {function} cb Callback to fire once stored auth credentials have been
+ *                      cleared. This is primarily for testing.
  * @return {Promise}
  */
-function logout() {
+function logout(cb) {
```
Comment (dylancwood) [inline scripts/lib/auth.js]: I see that the callback is used in tests, but why can't the test use promises instead of a callback?

Comment (dylancwood) [inline scripts/lib/auth.js]: What if we just return `jQuery.ajax`?

File: scripts/lib/auth.js (hunk 2 above)
Comment (swashcap) [inline scripts/lib/auth.js:97]: When testing it’s `localhost`.

File: scripts/lib/auth.js (hunk 3 above)
Comment (swashcap) [inline scripts/lib/auth.js]: Ah, good call. Didn’t know about that.

Comment (swashcap) [inline scripts/lib/auth.js]: Fixed in 9104157fe2022a8bc8dee78f1516216c2934418c.

File: scripts/lib/auth.js (hunk 4 above)
Comment (swashcap) [inline scripts/lib/auth.js]: 462f9d7d8fa9af60c49d7e9a633327e32d4ef2a8 addresses this.

File: scripts/lib/auth.js (hunk 2 above)
Comment (dylancwood) [inline scripts/lib/auth.js:97]: doh! Good point. Sorry.

**Pull request merged**